File: mobile/src/main/java/treehou/se/habit/gcm/GoogleCloudMessageConnector.java
package treehou.se.habit.gcm;

import android.content.Context;
import android.os.Build;
import android.util.Log;
import android.util.Pair;

import com.google.android.gms.ads.identifier.AdvertisingIdClient;
import com.google.android.gms.common.GooglePlayServicesNotAvailableException;
import com.google.android.gms.common.GooglePlayServicesRepairableException;
import com.google.android.gms.gcm.GoogleCloudMessaging;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.concurrent.Callable;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.functions.BiFunction;
import io.reactivex.functions.Function;
import io.reactivex.functions.Predicate;
import io.reactivex.schedulers.Schedulers;
import io.realm.Realm;
import se.treehou.ng.ohcommunicator.connector.models.OHServer;
import treehou.se.habit.module.ServerLoaderFactory;
import treehou.se.habit.util.ConnectionFactory;
import treehou.se.habit.util.Constants;

public class GoogleCloudMessageConnector {
    private static final String TAG = GoogleCloudMessageConnector.class.getSimpleName();

    @Inject
    protected ServerLoaderFactory serverLoaderFactory;
    @Inject
    protected Realm realm;
    @Inject
    protected ConnectionFactory connectionFactory;

    @Inject
    public GoogleCloudMessageConnector() {
    }

    public void registerGcm(Context context) {
        Observable.combineLatest(
                Observable.fromCallable(() -> {
                    try {
                        return AdvertisingIdClient.getAdvertisingIdInfo(context).getId();
                    } catch (IOException | GooglePlayServicesRepairableException | GooglePlayServicesNotAvailableException e) {
                        Log.e(TAG, "Error getting advertising ID: " + e.getMessage(), e);
                        return "";
                    }
                }).subscribeOn(Schedulers.io()),
                realm.asFlowable().toObservable().compose(serverLoaderFactory.loadServersRx()),
                Pair::create)
                .filter(deviceIdOHServerPair -> !deviceIdOHServerPair.first.isEmpty())
                .flatMap((Function<Pair<String, OHServer>, ObservableSource<?>>) deviceIdOHServerPair -> {

                    GoogleCloudMessaging gcm = GoogleCloudMessaging.getInstance(context);
                    String deviceId = deviceIdOHServerPair.first;
                    OHServer ohServer = deviceIdOHServerPair.second;
                    String registrationId;
                    try {
                        registrationId = gcm.register(Constants.GCM_SENDER_ID);
                    } catch (IOException e) {
                        Log.e(TAG, "Error getting GCM ID: " + e.getMessage(), e);
                        return Observable.empty();
                    }

                    String deviceModel;
                    try {
                        deviceModel = URLEncoder.encode(Build.MODEL, "UTF-8");
                    } catch (UnsupportedEncodingException ex) {
                        Log.d(TAG, "Could not encode device model: " + ex.getMessage());
                        return Observable.empty();
                    }

                    return connectionFactory.createServerHandler(ohServer, context).registerGcm(deviceId, deviceModel, registrationId);
                })
                .subscribe();
    }
}

File: mobile/src/main/java/treehou/se/habit/gcm/GoogleCloudMessageConnector.kt
package treehou.se.habit.gcm

import android.content.Context
import android.os.Build
import android.util.Log
import com.google.android.gms.ads.identifier.AdvertisingIdClient
import com.google.android.gms.gcm.GoogleCloudMessaging
import io.reactivex.Observable
import io.reactivex.functions.BiFunction
import io.reactivex.schedulers.Schedulers
import io.realm.Realm
import se.treehou.ng.ohcommunicator.connector.models.OHServer
import treehou.se.habit.dagger.ServerLoaderFactory
import treehou.se.habit.util.ConnectionFactory
import treehou.se.habit.util.Constants
import treehou.se.habit.util.RxUtil
import treehou.se.habit.util.logging.Logger
import java.net.URLEncoder
import javax.inject.Inject

class GoogleCloudMessageConnector @Inject
constructor() {

    @Inject lateinit var serverLoaderFactory: ServerLoaderFactory
    @Inject lateinit var realm: Realm
    @Inject lateinit var connectionFactory: ConnectionFactory
    @Inject lateinit var logger: Logger

    fun registerGcm(context: Context) {
        Observable.combineLatest(
                Observable.fromCallable<String> { AdvertisingIdClient.getAdvertisingIdInfo(context).id }.subscribeOn(Schedulers.io()),
                realm.asFlowable().toObservable().compose(serverLoaderFactory.loadServersRx()).compose(RxUtil().filterMyOpenhabServers()),
                BiFunction<String, OHServer, Pair<String, OHServer>> { deviceId, ohServer -> Pair(deviceId, ohServer) })
                .filter { deviceIdOHServerPair -> !deviceIdOHServerPair.first.isEmpty()}
                .distinct( {(_, server) -> (server.username.hashCode() + server.password.hashCode()) } )
                .flatMap({ (deviceId, ohServer) ->
                    registerGcmForClient(context, deviceId, ohServer);
                })
                .subscribe({},{ logger.e(TAG, "Failed to register GCM", it) })
    }

    /**
     * Filter myopenhab servers
     * @return remove all non myopenhabservers from stream
     */
    private fun registerGcmForClient(context: Context, deviceId: String, ohServer: OHServer): Observable<String> {
        try {
            Log.d(TAG, "Register GCM for user ${ohServer.username}")
            val gcm = GoogleCloudMessaging.getInstance(context)
            val registrationId = gcm.register(Constants.GCM_SENDER_ID)
            val deviceModel = URLEncoder.encode(Build.MODEL, "UTF-8")
            return connectionFactory.createServerHandler(ohServer, context).registerGcm(deviceId, deviceModel, registrationId)
        }catch (_: Exception){
            return Observable.empty<String>()
        }
    }

    companion object {
        val TAG =  GoogleCloudMessageConnector::class.java.simpleName;
    }
}
